modules/Bio/EnsEMBL/Variation/Pipeline/AncestralAlleles/Assign.pm

```diff
@@ -0,0 +1,168 @@
+}
+
+sub dump_variation_features {
+  my $self = shift;
+  my $species_name = $self->param('species_name');
+  my $species_dir = $self->param('species_dir');
+  my $batch = $self->param('batch'); 
+  my ($start, $end) = ($batch->{start}, $batch->{end});
+  my $vf_file = "$species_dir/vf_$start\_$end";
+  my $aa_file = "$species_dir/aa_$start\_$end";
+  $self->param('vf_file', $vf_file);
+  $self->param('aa_file', $aa_file);
+  if (! -e "$vf_file") {
+    my $fh = FileHandle->new($vf_file, 'w');
+    $self->param('species', $species_name);
+    my $vdba = $self->get_species_adaptor('variation');
+    my $dbc = $vdba->dbc;
+    my $sth = $dbc->prepare(qq{
+      SELECT vf.variation_feature_id, sr.name, vf.seq_region_start, vf.seq_region_end, vf.seq_region_strand, vf.ancestral_allele
+      FROM variation_feature vf, seq_region sr
+      WHERE vf.seq_region_id = sr.seq_region_id
+      AND vf.variation_feature_id >= ?
+      AND vf.variation_feature_id <= ?
+      ORDER BY vf.variation_feature_id;}, {mysql_use_result => 1}
+    ) or die $dbc->errstr;
+    $sth->execute($start, $end) or die $dbc->errstr;
+    my ($vf_id, $seq_region, $start, $end, $strand, $aa);
+    $sth->bind_columns(\($vf_id, $seq_region, $start, $end, $strand, $aa));
+    while ($sth->fetch) {
+      $aa ||= 'NULL';
+      print $fh join("\t", ($vf_id, $seq_region, $start, $end, $strand, $aa)), "\n";
+    }
+    $sth->finish();
+    $fh->close();
+  }
+}
+
+sub assign_ancestral_alleles {
+  my $self = shift;
+  my $fasta_files_dir = $self->param('fasta_files_dir');
+  my $db = Bio::DB::Fasta->new($fasta_files_dir, -reindex => 1);
+  my @sequence_ids = $db->get_all_ids;
+  my %sequence_id_2_chr_number;
+  foreach my $sequence_id (@sequence_ids) {
+    my @split = split(/:/, $sequence_id);
+    $sequence_id_2_chr_number{$split[2]} = $sequence_id;
+  }
+
+  my $vf_file = $self->param('vf_file');
+  my $aa_file = $self->param('aa_file');
+  my $input = FileHandle->new($vf_file, 'r');
+  my $fh    = FileHandle->new("$aa_file.out", 'w');
+  my $err   = FileHandle->new("$aa_file.err", 'w');
+  while (<$input>) {
+    chomp;
+    my ($vf_id, $chrom, $start, $end, $strand, $old_aa) = split /\t/;
+    my $set_aa_to_null = 1;
+    if ($start > $end) {
+      print $err "Insertion $vf_id\n";
+    }
+    elsif (($end - $start) > 50) {
+      print $err "Longer than 50bp $vf_id\n";
+    } else {
+      my $chrom_name = $sequence_id_2_chr_number{$chrom};
+      if (!($chrom_name && $start && $end)) {
+        print $err "Incomplete coords $vf_id\n";
+      } else {
+        my $aa = $db->seq("$chrom_name:$start,$end");
+        if ($aa && $aa =~ m/^[ACGTacgt]+$/) {
+          $aa = uc $aa;
+          $set_aa_to_null = 0;
+          print $fh "UPDATE variation_feature SET ancestral_allele='$aa' WHERE variation_feature_id=$vf_id;\n" if ($aa ne $old_aa);
+        } else {
+          print $err "No AA in fasta $vf_id\n";
+        }
+      }
+    }
+    if ($set_aa_to_null && $old_aa ne 'NULL') {
```

Comment:
It would be useful to check that the entire batch is not being set to null.

Reply:
Do you think this would mean something is wrong with the ancestral FASTA file or that particular job has problems reading from disk? Usually it will flag up in the final count comparison if the number of NULLs has gone up. But it is true that it will be difficult tell if there was a systematic error.

Reply:
I'm thinking about particular jobs having problems reading from disk. I know we shouldn't see this.